### PR TITLE
Add Functionalities

### PR DESCRIPTION
**Pull request by kristofbolyai**
This pull request introduces the "functionality" base, filling the gap we currently had with features. 

- Functionalities can be used to add event dependent methods, which can be registered or unregistered by other code (like commands). Basically, they are features, but they can not be enabled or disabled, as doing so does not make sense.
- They can be also used for adding any code that should not be in an "util" class, and currently do not have a proper place (like NPE packet fixes in mixin code).
- If any mixins are added that are not just event calls, but inject more code, a functionality with a static method should be called in the mixin. (Writing this up made me think, should `WynncraftButtonFeature` be a feature, or a functionality? I see it more a functionality, as disabling it does unregister the button, although it can be argued that it's just an incomplete feature)

### CODE REVIEW

**Comment by magicus** [comment]
Hm, you're making functionalities more rigid than I was envisioning them. I was more thinking like have a collection of classes with not really a strict "per module" division like features. A more fluid approach, depending on the needs. My thinking was that the "functionalities" were like the odds bits and pieces that did not really fit into the strict feature model. Is the idea with the registry to help functionalities register themselves on the event bus?

**Comment by kristofbolyai** [comment]
> Hm, you're making functionalities more rigid than I was envisioning them. I was more thinking like have a collection of classes with not really a strict "per module" division like features. A more fluid approach, depending on the needs. My thinking was that the "functionalities" were like the odds bits and pieces that did not really fit into the strict feature model. Is the idea with the registry to help functionalities register themselves on the event bus?

The registry is there for the crash log extension, and we need to instantiate the functionality classes somewhere. Since registering is a static method, it is just a simple call. 

Also, I am all for making them this separated, as I fear if we don't, we end up creating legacy modules' ClientEvents mess.

File: common/src/main/java/com/wynntils/core/functionalities/Functionality.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            throw new IllegalStateException(
+                    "Functionality cannot be unregistered as it is not registered to event bus.");
+
+        registeredToEventBus = false;
+        WynntilsMod.getEventBus().unregister(this.getClass());
+    }
+
+    public static Functionality getInstance() {
+        return INSTANCE;
+    }
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/core/functionalities/Functionality.java:52]: i don't think this singleton impl really works

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/functionalities/Functionality.java:52]: pretty sure the sub classes all point at the same instance

**Comment by Incompleteusern** [comment]
I don't see how functionalities are fundamentally different from features, other than that they are enabled and disabled either not at all or externally. To me it seems a functionality could be implemented as a feature using the Condition system. If they can't be enabled or disabled in general then it ends up being a superclass of Feature.

File: common/src/main/java/com/wynntils/core/functionalities/Functionality.java (hunk 1 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/core/functionalities/Functionality.java:52]: Actually that might be true, singletons with abstract bases are strange

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/core/functionalities/Functionality.java:52]: I will refactor

**Comment by Incompleteusern** [comment]
Also, functionality is a long, awkward word, and already has a definition

**Comment by kristofbolyai** [comment]
> I don't see how functionalities are fundamentally different from features, other than that they are enabled and disabled either not at all or externally. To me it seems a functionality could be implemented as a feature using the Condition system. If they can't be enabled or disabled in general then it ends up being a superclass of Feature.

What condition would you add?

I see features something you can turn on or off, and modify its settings. Functionalities don't have a clear on/off state, and thus, shouldn't be controlled by a player (although I can see some functionalities have settings).

**Comment by kristofbolyai** [comment]
> Also, functionality is a long, awkward word, and already has a definition

Do you have any suggestions for a better name?

**Comment by Incompleteusern** [comment]
For lootruns some condition that is dependent on `LootRunUtils`. Maybe even a `ManagedFeature` which has explicit support for an external enabling.

Features having `enable` and `disable` doesn't mean that the player can turn therefore them on or off, its more of a way to just in general control whether a piece of code is on or off, which allows for /wynntils reload.

**Comment by kristofbolyai** [comment]
> For lootruns some condition that is dependent on `LootRunUtils`. Maybe even a `ManagedFeature` which has explicit support for an external enabling.
> 
> 
> 
> Features having `enable` and `disable` doesn't mean that the player can turn therefore them on or off, its more of a way to just in general control whether a piece of code is on or off, which allows for /wynntils reload.

If we add ManagedFeature, what does reloading `LootrunFeature` do?

**Comment by Incompleteusern** [comment]
To be honest `/wynntils reload` should only reload features that have the potential to fail on enabling. Reloading `LootrunFeature` likely has no purpose to be fair. 

I think you raise a good point about how not all features have a clear on/off reason, but I don't think functionalities is a way to fix it. I think some superclass without enabling/disabling might also work

**Comment by kristofbolyai** [comment]
> To be honest `/wynntils reload` should only reload features that have the potential to fail on enabling. Reloading `LootrunFeature` likely has no purpose to be fair. 
> 
> 
> 
> I think you raise a good point about how not all features have a clear on/off reason, but I don't think functionalities is a way to fix it. I think some superclass without enabling/disabling might also work

Can you share a design sketch of how you would go about making the superclass?

Let's all come to an agreement on this before making 10 different designs.

**Comment by Incompleteusern** [comment]
```
FeatureBase/Feature (TBH these 2 could just be merged, internally has enabling/disabling but not available to player)
 |     |    \ > Features which are always enabled 
 V    \> ManagableFeature (Enabled/disabled externally like a lootrun)
EnableableFeature (Player can change whether its enabled)
```
Thought about it some more and am no longer convinced in a superclass.

`\wynntils reload` is handled only for api conditions, which should also be better supported.

**Comment by kristofbolyai** [comment]
> ```
> 
> FeatureBase/Feature (TBH these 2 could just be merged, internally has enabling/disabling but not available to player)
> 
>  |     |    \ > Features which are always enabled 
> 
>  V    \> ManagableFeature (Enabled/disabled externally like a lootrun)
> 
> EnableableFeature (Player can change whether its enabled)
> 
> ```
> 
> Thought about it some more and am no longer convinced in a superclass.
> 
> 
> 
> `\wynntils reload` is handled only for api conditions, which should also be better supported.
> 
> 

Design seems nice (although maybe we should think of better names).

**Comment by magicus** [comment]
Can you give me a day or two to think about it? I feel I have opinions but not really time right now.

**Comment by P0keDev** [comment]
> ```
> FeatureBase/Feature (TBH these 2 could just be merged, internally has enabling/disabling but not available to player)
>  |     |    \ > Features which are always enabled 
>  V    \> ManagableFeature (Enabled/disabled externally like a lootrun)
> EnableableFeature (Player can change whether its enabled)
> ```
> 
> Thought about it some more and am no longer convinced in a superclass.
> 
> `\wynntils reload` is handled only for api conditions, which should also be better supported.

i agree that `FeatureBase` and `Feature` can probably be merged into one thing. i also don't think it's necessary to distinguish between features that are always on and features that are toggled by other code - the former could just enable itself, for example, either through its `init` function or maybe an annotation if we want to be fancy

as for name ideas: `InternalFeature` and `ExternalFeature`? maybe `UserFeature` for the latter?

**Comment by kristofbolyai** [comment]
@magicus Have you made up your mind?

**Comment by magicus** [comment]
No, sorry, I don't have time for Wynntils right now. :( You'll have to go ahead and do what you think is best. If I'm extremely unsatisfied I might come back later and suggest some alternative approach, if I can find one. :-)

**Comment by kristofbolyai** [comment]
Superseded by #77

**Pull request closed without merging**